annotations/src/main/java/com/uber/nullaway/annotations/EnsuresNonNullIf.java

```diff
@@ -0,0 +1,40 @@
+package com.uber.nullaway.annotations;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * An annotation describing a nullability post-condition for an instance method. Each parameter to
+ * the annotation should be a field of the enclosing class. The method must ensure that the method
+ * returns true in case the fields are non-null. The method can also return true in case the fields
+ * are null (inverse logic), and in such case, you must set trueIfNonNull to false. NullAway
+ * verifies that the property holds.
+ *
+ * <p>Here is an example:
+ *
+ * <pre>
+ * class Foo {
+ *     {@literal @}Nullable Object theField;
+ *     {@literal @}EnsuresNonNullIf("theField") // @EnsuresNonNullIf("this.theField") is also valid
+ *     boolean hasTheField() {
+ *         return theField != null;
+ *     }
+ *     void bar() {
+ *         if(!hasTheField()) {
+ *             return;
+ *         }
+ *         // No error, NullAway knows theField is non-null after call to hasTheField()
+ *         theField.toString();
+ *     }
+ * }
+ * </pre>
+ */
+@Retention(RetentionPolicy.CLASS)
+@Target({ElementType.METHOD})
+public @interface EnsuresNonNullIf {
+  String[] value();
+
+  boolean trueIfNonNull() default true;
```

Comment:
I made it optional because my gut feeling says that the large majority of such methods are in the form of `hasX` rather than `doesNotHaveX`, which is more exceptional. I am more of a fan of sensible defaults instead of forcing people to explicit mark every decision, but of course, whatever you say here!

I don't get your comment about the "if" in the name. Should we change the name to something else? `trueIfNonNull` was just a placeholder as I couldn't think of something better!

Reply:
> I made it optional because my gut feeling says that the large majority of such methods are in the form of `hasX` rather than `doesNotHaveX`, which is more exceptional. I am more of a fan of sensible defaults instead of forcing people to explicit mark every decision, but of course, whatever you say here!

For the `false` case I was thinking of methods like `StringUtils.isEmpty(s)` which ensures `s` is `@NonNull` if it returns `false`.  You can see a list of similar methods [here](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java#L695-L727).  There might be analogous methods on fields.  I think you may be right that the other case is more common, but I'm not sure.

> I don't get your comment about the "if" in the name. Should we change the name to something else? `trueIfNonNull` was just a placeholder as I couldn't think of something better!

Sorry, I meant the `If` in `EnsuresNonNullIf`.  To me, given the annotation name, it is natural to want to specify the condition of the `If` part to be true or false.

Overall, I'd rather stick to doing something similar to the Checker Framework here.  If we really want to simplify the defaults, I'd rather have a separate annotation like `@EnsuresNonNullIfTrue`; but I feel that it's overkill.